fl4health/clients/fed_pca_client.py

```diff
@@ -0,0 +1,125 @@
+
+        Returns:
+            NDArrays: packed principal components and their corresponding eigenvalues.
+        """
+        pcs, eigenvals = self._pca.get_pcs()
+        return self.pca_packer.pack(pcs, eigenvals)
+
+    def set_parameters(self, packed_pcs: NDArrays, config: Config) -> None:
+        """
+        Sets the merged principal components transfered from the server.
+
+        Args:
+            packed_pcs (NDArrays): principal components and their corresponding eigenvalues packed together.
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+        """
+        pcs, eigenvals = self.pca_packer.unpack(packed_pcs)
+        self._pca.update_pcs(pcs, eigenvals)
+        self._pca.save_pcs(self.pc_path)
+
+    def fit(self, parameters: NDArrays, config: Config) -> Tuple[NDArrays, int, Dict[str, Scalar]]:
+        """Compute PCA using the locally held dataset.
+
+        Parameters
+        ----------
+        parameters : NDArrays
+            This is unused in the case of PCA.
+        config : Dict[str, Scalar]
+            Configuration parameters.
+
+        Returns
+        -------
+        packed_pcs : NDArrays
+            Principal components of local dataset along with their explained variance.
+        num_examples : int
+            The number of examples used for training.
+        metrics : Dict[str, Scalar]
+            A dictionary mapping arbitrary string keys to values of type
+            bool, bytes, float, int, or str.
+        """
+        X = self.get_data_numpy()
+        principal_components, eigenvalues = self._pca.compute_pc(X)
+        packed_pcs = self.pca_packer.pack(principal_components, eigenvalues)
+        num_examples = X.shape[0]
+        return packed_pcs, num_examples, {}
+
+    def get_data_numpy(self) -> NDArray:
+        """
+        Return this client's data as a numpy array so PCA can be computed.
+
+        Raises:
+            NotImplementedError: This method should be implemented by the user
+            based on the specific dataset encountered.
+
+        Returns:
+            X (NDArray): This client's data as a numpy array.
+        """
+        raise NotImplementedError
+
+    def narrow_config_type(self, config: Config, config_key: str, narrow_type_to: Type[T]) -> T:
+        if config_key not in config:
+            raise ValueError(f"{config_key} is not present in the Config.")
+
+        config_value = config[config_key]
+        if isinstance(config_value, narrow_type_to):
+            return config_value
+        else:
+            raise ValueError(f"Provided configuration key ({config_key}) value does not have correct type")
+
+    def setup_client(self, config: Config) -> None:
+        self.initialized = True
+
+    def evaluate(self, parameters: NDArrays, config: Dict[str, Scalar]) -> Tuple[float, int, Dict[str, Scalar]]:
```

Comment:
One thing I'm not sure about is what the `evaluate` step should be for federated PCA... Since PCA does not fit into the normal training paradigm, and from what I've read the quality of PCA is usually evaluated through the downstream tasks.